### PR TITLE
Implement timestamp +/- IntervalYearMonth

### LINKED ISSUE #8665: Addition of substraction with interval Month and Year is not supported (Day is supported).
### Description

Following query works
`select DATE(ds) - INTERVAL '1' DAY from mytable;`
However, the addition/subtraction does not work for MONTH or YEAR interval. Following queries does not work:
`select DATE(ds) - INTERVAL '1' MONTH from mytable;`
`select DATE(ds) - INTERVAL '1' YEAR from mytable;`

Since `DAY` is already supported, completing it with other time units will unblock many queries with similar shapes.

### PR DESCRIPTION
**Pull request by xumingming**
Fixes #8665

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@xumingming I didn't see this PR before I started working on #8911. Would you update this PR to only add support for timestamp + interval?

**Comment by xumingming** [comment]
@mbasmanova Removed `Date` related code.

File: velox/expression/UdfTypeResolver.h [hunk 1]
```diff
@@ -129,6 +129,13 @@ struct resolver<IntervalDayTime> {
   using out_type = int64_t;
 };
 
+template <>
+struct resolver<IntervalYearMonth> {
```
Comment (mbasmanova) [inline velox/expression/UdfTypeResolver.h:133]: Is this needed? Would you try rebasing?

File: velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp [hunk 2]
```diff
@@ -850,6 +850,75 @@ TEST_F(DateTimeFunctionsTest, plusMinusDateIntervalDayTime) {
...
+
+  // Some special dates.
+  EXPECT_EQ("2001-06-30 04:05:06", plus("2001-05-31 04:05:06", 1));
+  EXPECT_EQ("2001-05-30 04:05:06", plus("2001-04-30 04:05:06", 1));
+  EXPECT_EQ("2001-02-28 04:05:06", plus("2001-01-31 04:05:06", 1));
+  EXPECT_EQ("2000-02-29 04:05:06", plus("2000-01-31 04:05:06", 1));
+  EXPECT_EQ("2000-02-29 04:05:06", plus("2000-01-29 04:05:06", 1));
+}
+
+TEST_F(DateTimeFunctionsTest, intervalYearMonthPlusTimestamp) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:899]: nit: perhaps, combine this test with the previous one by changing the lambda to run both t + i and i + t, assert that results are the same, then return the result.

**Review by mbasmanova — APPROVED** [review]
@xumingming Looks great. Thank you for adding these functions.

**Comment by xumingming** [comment]
@mbasmanova Rebased and addressed the comments.

**Comment by mbasmanova** [comment]
@xumingming CI is red. Would you rebase to see if failures go away?

**Comment by mbasmanova** [comment]
Gettiing UBSAN failures on fuzzer:

```
I0304 17:38:05.975989 292039 ExpressionFuzzerVerifier.cpp:308] ==============================> Started iteration 110 (seed: 2134270587)
I0304 17:38:06.011584 292039 ExpressionVerifier.cpp:51] Executing expression 0 : array_min(array_intersect(slice(flatten(remove_nulls(array_max("c0"))),"c1",bitwise_arithmetic_shift_right(month(minus("c2","c3")),"c4")),flatten(concat(null,"c5",array_sort(1 elements starting at 329 {[329->784] [784->226] 5 elements starting at 2121 {true, null, false, true, true}},lambda ROW<__a0:ARRAY<BOOLEAN>> -> null),12 elements starting at 81 {10 elements starting at 761 {null, false, false, null, true, ...}, 12 elements starting at 771 {true, false, null, true, false, ...}, 1 elements starting at 783 {false}, 7 elements starting at 784 {true, false, true, true, true, ...}, 5 elements starting at 791 {true, false, true, false, false}, ...},array_remove("c5",slice(flatten(remove_nulls(array_max("c0"))),"c1",bitwise_arithmetic_shift_right(month(minus("c2","c3")),"c4")))))))
buck-out/v2/gen/fbcode/0155eabfa767915a/velox/external/date/__velox_external_date__/buck-headers/velox/external/date/date.h:2765:21: runtime error: signed integer overflow: -106833 * 146097 cannot be represented in type 'int'
    #0 0x7fc4b643be9c in facebook::velox::date::year_month_day::to_days() const fbcode/velox/external/date/date.h:2765
    #1 0x7fc4b6438afa in facebook::velox::date::year_month_day::operator std::chrono::time_point<std::chrono::_V2::system_clock, std::chrono::duration<int, std::ratio<86400l, 1l>>>() const fbcode/velox/external/date/date.h:2772
    #2 0x7fc4b6787a3a in addToDate fbcode/velox/functions/prestosql/DateTimeImpl.h:126
    #3 0x7fc4b6787a3a in addToTimestamp fbcode/velox/functions/prestosql/DateTimeImpl.h:157
    #4 0x7fc4b6787a3a in call fbcode/velox/functions/prestosql/DateTimeFunctions.h:500
    #5 0x7fc4b6787a3a in callImpl fbcode/velox/core/SimpleFunctionMetadata.h:780
    #6 0x7fc4b6787a3a in call fbcode/velox/core/SimpleFunctionMetadata.h:720
    #7 0x7fc4b6787a3a in doApplyNotNull<2UL, facebook::velox::Timestamp, int, 0> fbcode/velox/expression/SimpleFunctionAdapter.h:843
    #8 0x7fc4b6787a3a in doApplyNotNull<1UL, const facebook::velox::exec::VectorReader<facebook::velox::IntervalYearMonth>, facebook::velox::Timestamp, 0> fbcode/velox/expression/SimpleFunctionAdapter.h:833
    #9 0x7fc4b6787a3a in doApplyNotNull<0UL, facebook::velox::exec::VectorReader<facebook::velox::Timestamp>, facebook::velox::exec::VectorReader<facebook::velox::IntervalYearMonth>, 0> fbcode/velox/expression/SimpleFunctionAdapter.h:833
    #10 0x7fc4b6787a3a in operator()<int> fbcode/velox/expression/SimpleFunctionAdapter.h:653
    #11 0x7fc4b6787a3a in operator()<int> fbcode/velox/expression/EvalCtx.h:105
```

and AddressSanitizer: heap-use-after-free for the unit test

```
[ RUN      ] DateTimeFunctionsTest.timestampMinusIntervalYearMonth

=================================================================
==291929==ERROR: AddressSanitizer: heap-use-after-free on address 0x62f00000e444 at pc 0x0000027fbf7c bp 0x7ffc9294c3a0 sp 0x7ffc9294bb60
READ of size 15 at 0x62f00000e444 thread T0
SCARINESS: 54 (multi-byte-read-heap-use-after-free)
    #0 0x27fbf7b in MemcmpInterceptorCommon(void*, int (*)(void const*, void const*, unsigned long), void const*, void const*, unsigned long) ubsan.c
    #1 0x27fc538 in __interceptor_memcmp (/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/0155eabfa767915a/velox/functions/prestosql/tests/__velox_functions_test__/velox_functions_test+0x27fc538)
    #2 0xec87f0 in facebook::velox::StringView::operator==(facebook::velox::StringView const&) const fbcode/velox/type/StringView.h:146
    #3 0x1d47567 in std::enable_if<is_convertible<decltype(std::declval<char const (&) [20]>() == std::declval<facebook::velox::StringView const&>()), bool>::value, bool>::type std::operator==<facebook::velox::StringView, char [20]>(char const (&) [20], std::optional<facebook::velox::StringView> const&) fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/optional:1174
    #4 0x1d47403 in testing::AssertionResult testing::internal::CmpHelperEQ<char [20], std::optional<facebook::velox::StringView>>(char const*, char const*, char const (&) [20], std::optional<facebook::velox::StringView> const&) fbsource/gtest/gtest.h:1379
    #5 0x1d46307 in testing::AssertionResult testing::internal::EqHelper::Compare<char [20], std::optional<facebook::velox::StringView>, (void*)0>(char const*, char const*, char const (&) [20], std::optional<facebook::velox::StringView> const&) fbsource/gtest/gtest.h:1398
    #6 0x1bc6e1c in DateTimeFunctionsTest_timestampMinusIntervalYearMonth_Test::TestBody() fbcode/velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp:895
...
SUMMARY: AddressSanitizer: heap-use-after-free ubsan.c in MemcmpInterceptorCommon(void*, int (*)(void const*, void const*, unsigned long), void const*, void const*, unsigned long)
Shadow bytes around the buggy address:
  0x0c5e7fff9c30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c5e7fff9c40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c5e7fff9c50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c5e7fff9c60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c5e7fff9c70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c5e7fff9c80: 00 00 00 00 00 00 00 00[00]00 00 00 00 00 00 00
  0x0c5e7fff9c90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c5e7fff9ca0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c5e7fff9cb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c5e7fff9cc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c5e7fff9cd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==291929==ABORTING
```

**Pull request closed without merging**